src/main/java/com/google/sps/data/Journey.java

```diff
@@ -79,45 +87,184 @@ public int getNumberOfWaypoints() {
+   * is lastWaypoint; config is a bitmask where bits set to 1 represent currently visited waypoints
+   */
+  private ArrayList<ArrayList<Pair>> initializeBestTimeMatrix() throws ApiException, IOException, InterruptedException, DataNotFoundException {
+    ArrayList<ArrayList<Pair>> bestTime = new ArrayList<>();
+    int numberOfWaypoints = getNumberOfWaypoints();
+    for (int config = 0; config <= (1 << numberOfWaypoints); ++config) {
+      // add the config-th line to the matrix with numberOfWaypoints columns of Integer.MAX_VALUE
+      ArrayList<Pair> currentRow = new ArrayList<Pair>();
+
+      for (int waypointIndex = 0; waypointIndex <= numberOfWaypoints; ++ waypointIndex) {
+        currentRow.add(new Pair(Integer.MAX_VALUE, -1));
+      }
+      bestTime.add(currentRow);
+    }
+
+    int waypointIndex = 0;
+    for (Point waypoint : waypoints) {
+      Pair currentElement = bestTime.get(1 << waypointIndex).get(waypointIndex);
+      currentElement.first = pathFinder.getTimeInSecondsBetweenPoints(start, waypoint);
+      waypointIndex += 1;
+    }
+
+    return bestTime;
+  }
+
+  /**
+   * Computes and returns the bestTime matrix such that bestTime[config][latsWaypoint] is the min
+   * time needed to visit all points with bits set to 1 in config such that the path ends in point
+   * lastWaypoint.
+   */
+  public ArrayList<ArrayList<Pair>> findOptimalTimePaths()  throws ApiException, DataNotFoundException, IOException, InterruptedException {
+    ArrayList<ArrayList<Pair>> bestTime = initializeBestTimeMatrix();
+    ArrayList<Integer> pointsAfter = getPointsAfterFromRestricitions();
+    int numberOfWaypoints = getNumberOfWaypoints();
+    int allWaypointsConfig = (1 << numberOfWaypoints) - 1;
+    // start with config = 1 so that at least 1 point is set to 1 in config
+    for (int config = 1; config < allWaypointsConfig; ++config) {
+      for (int waypointIndex = 0; waypointIndex < numberOfWaypoints; ++ waypointIndex) {
+        // check for waypointIndex, the next unvisited point for which none of the points that must
+        // be visited after him have been visited.
+        // TODO[ak47na]: add pointsBefore array and check that all points that must be visited
+        // before the current point have already been visited.
+        if ((config & (1 << waypointIndex)) == 0 && ((pointsAfter.get(waypointIndex) & config) == 0)) {
+          // Set bit wayPointIndex to 1 in newConfig if waypointIndex is set to zero in config AND
+          // all restrictions for waypointIndex are satisfied
+          int newConfig = config | (1 << waypointIndex);
+          for (int lastWaypoint = 0; lastWaypoint < numberOfWaypoints; ++lastWaypoint) {
+            // the lastWaypoint visited in config must be set to 1
+            if ((config & (1 << lastWaypoint)) != 0) {
+              // add the time needed to get from lastWaypoint to waypointIndex to bestTime[config][lastWaypoint]
+              // TODO[ak47na]: handle case when
```

Comment:
?

Reply:
it was a TODO to handle the case when `getTimeInSecondsBetweenPoints` fails, but know it throws `ApiException`. I removed the comment.